File: scripts/generate_inline_routes.py
from pathlib import Path
import json
import re


ROOT = Path("/Users/betus/Documents/trae_projects/baby-grow-mini")
PROTOTYPE_ROOT = ROOT / "prototype_demo"
ROUTES_PATH = PROTOTYPE_ROOT / "routes.js"


def collect_route_sources():
    routes_text = ROUTES_PATH.read_text(encoding="utf-8")
    pattern = re.compile(r'id:\s*"([^"]+)"[\s\S]*?src:\s*"([^"]+)"', re.M)
    pairs = pattern.findall(routes_text)
    inline = {}
    for route_id, src in pairs:
        abs_path = (PROTOTYPE_ROOT / src).resolve()
        inline[route_id] = abs_path.read_text(encoding="utf-8")
    return inline


def write_inline_routes_file(inline):
    out_path = PROTOTYPE_ROOT / "inline-routes.js"
    out_path.write_text(
        "window.PROTOTYPE_INLINE_HTML = " + json.dumps(inline, ensure_ascii=False) + ";\n",
        encoding="utf-8"
    )


def patch_entry_files():
    old = '<script src="./routes.js"></script>\n  <script src="./app.js"></script>'
    new = '<script src="./routes.js"></script>\n  <script src="./inline-routes.js"></script>\n  <script src="./app.js"></script>'
    updated = 0
    for path in PROTOTYPE_ROOT.glob("*.html"):
        html = path.read_text(encoding="utf-8")
        if old in html:
            path.write_text(html.replace(old, new), encoding="utf-8")
            updated += 1
    return updated


def main():
    inline = collect_route_sources()
    write_inline_routes_file(inline)
    updated = patch_entry_files()
    print(f"routes={len(inline)}")
    print(f"updated={updated}")


if __name__ == "__main__":
    main()

File: scripts/generate_inline_routes.js
const fs = require("fs");
const path = require("path");
const vm = require("vm");

const ROOT = "/Users/betus/Documents/trae_projects/baby-grow-mini";
const PROTOTYPE_ROOT = path.join(ROOT, "prototype_demo");
const ROUTES_PATH = path.join(PROTOTYPE_ROOT, "routes.js");

function readPrototypeData() {
  const code = fs.readFileSync(ROUTES_PATH, "utf8");
  const sandbox = { window: {} };
  vm.createContext(sandbox);
  vm.runInContext(code, sandbox);
  if (!sandbox.window.PROTOTYPE_DATA || !Array.isArray(sandbox.window.PROTOTYPE_DATA.routes)) {
    throw new Error("Failed to parse routes.js");
  }
  return sandbox.window.PROTOTYPE_DATA;
}

function writeInlineRoutes(routes) {
  const inlineHtml = {};
  routes.forEach((route) => {
    const absPath = path.resolve(PROTOTYPE_ROOT, route.src);
    inlineHtml[route.id] = fs.readFileSync(absPath, "utf8");
  });

  const outPath = path.join(PROTOTYPE_ROOT, "inline-routes.js");
  fs.writeFileSync(
    outPath,
    "window.PROTOTYPE_INLINE_HTML = " + JSON.stringify(inlineHtml) + ";\n",
    "utf8"
  );

  return Object.keys(inlineHtml).length;
}

function patchEntries() {
  const oldSnippet = '<script src="./routes.js"></script>\n  <script src="./app.js"></script>';
  const newSnippet = '<script src="./routes.js"></script>\n  <script src="./inline-routes.js"></script>\n  <script src="./app.js"></script>';
  let updated = 0;

  fs.readdirSync(PROTOTYPE_ROOT).forEach((name) => {
    if (!name.endsWith(".html")) {
      return;
    }
    const filePath = path.join(PROTOTYPE_ROOT, name);
    const html = fs.readFileSync(filePath, "utf8");
    if (html.includes(oldSnippet)) {
      fs.writeFileSync(filePath, html.replace(oldSnippet, newSnippet), "utf8");
      updated += 1;
    }
  });

  return updated;
}

function main() {
  const data = readPrototypeData();
  const routesCount = writeInlineRoutes(data.routes);
  const updatedCount = patchEntries();
  console.log(`routes=${routesCount}`);
  console.log(`updated=${updatedCount}`);
}

main();
